Implement trusted types enforcement on Function constructor

lukewarlow opened the pull request:
#### 152e920a5ac26047b9103bd750367343205e7752
<pre>
Implement trusted types enforcement on Function constructor
<a href="https://bugs.webkit.org/show_bug.cgi?id=273187">https://bugs.webkit.org/show_bug.cgi?id=273187</a>

Reviewed by Yusuke Suzuki.

This patch adds trusted types enforcement to the Function constructor as well as updating the eval implementation.

The canCompileStrings global method table function no longer takes a JSValue argument,
this is because it is now only called for untrusted input from eval.

The implementation of TT enforcement for both direct and indirect eval is updated to do more work in JSC.
The structure of the TrustedScript type is used by JSC to determine if an object should be evaluated,
rather than always calling codeForEval. Only if the structures don&apos;t match is codeForEval called, this can happen
if someone changes the instance properties such as for polyfills.

The canCompileStrings call is now only done if the input is known
to be untrusted (raw string rather than TrustedScript argument).

The Function constructor is now also updated such that when TT enforcement is enabled through CSP, similar logic
comparing the arguments to the TrustedScript structure is used.
If not all of the arguments match the structure then they&apos;re not trusted so we fallback to calling canCompileStrings,
with a new ArgList atgument.

The ArgList is used to check if the arguments are modified trusted script objects, which are accepted providing the
stringifier isn&apos;t modified.

DebuggerEvalEnabler is also updated to disable trusted types and re-enable it, so that
web inspector can continue working on sites with TT enforced.

* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/reporting/report-clips-sample.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-csp-tt-default-policy-mutate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-csp-tt-no-default-policy-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-function-constructor-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-function-constructor.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-with-permissive-csp-expected.txt:
* Source/JavaScriptCore/API/JSAPIGlobalObject.cpp:
(JSC::JSAPIGlobalObject::globalObjectMethodTable):
* Source/JavaScriptCore/API/JSAPIGlobalObject.mm:
(JSC::JSAPIGlobalObject::globalObjectMethodTable):
* Source/JavaScriptCore/debugger/DebuggerEvalEnabler.h:
(JSC::DebuggerEvalEnabler::DebuggerEvalEnabler):
(JSC::DebuggerEvalEnabler::~DebuggerEvalEnabler):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::eval):
* Source/JavaScriptCore/jsc.cpp:
* Source/JavaScriptCore/runtime/DirectEvalExecutable.cpp:
(JSC::DirectEvalExecutable::create):
* Source/JavaScriptCore/runtime/FunctionConstructor.cpp:
(JSC::constructFunction):
* Source/JavaScriptCore/runtime/GlobalObjectMethodTable.h:
* Source/JavaScriptCore/runtime/IndirectEvalExecutable.cpp:
(JSC::IndirectEvalExecutable::createImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::baseGlobalObjectMethodTable):
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::arrayIteratorStructure const):
(JSC::JSGlobalObject::trustedScriptStructure):
(JSC::JSGlobalObject::reportViolationForUnsafeEval):
(JSC::JSGlobalObject::canCompileStrings):
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WebCore/bindings/js/JSDOMGlobalObject.cpp:
(WebCore::JSDOMGlobalObject::codeForEval):
(WebCore::JSDOMGlobalObject::canCompileStrings):
(WebCore::JSDOMGlobalObject::trustedScriptStructure):
* Source/WebCore/bindings/js/JSDOMGlobalObject.h:
* Source/WebCore/bindings/js/JSDOMWindowBase.cpp:
(WebCore::JSDOMWindowBase::globalObjectMethodTable):
(WebCore::JSDOMWindowBase::reportViolationForUnsafeEval):
(WebCore::JSDOMWindowBase::codeForEval): Deleted.
(WebCore::JSDOMWindowBase::canCompileStrings): Deleted.
* Source/WebCore/bindings/js/JSDOMWindowBase.h:
* Source/WebCore/bindings/js/JSShadowRealmGlobalScopeBase.cpp:
(WebCore::JSShadowRealmGlobalScopeBase::globalObjectMethodTable):
(WebCore::JSShadowRealmGlobalScopeBase::reportViolationForUnsafeEval):
(WebCore::JSShadowRealmGlobalScopeBase::codeForEval): Deleted.
(WebCore::JSShadowRealmGlobalScopeBase::canCompileStrings): Deleted.
* Source/WebCore/bindings/js/JSShadowRealmGlobalScopeBase.h:
* Source/WebCore/bindings/js/JSWorkerGlobalScopeBase.cpp:
(WebCore::JSWorkerGlobalScopeBase::globalObjectMethodTable):
(WebCore::JSWorkerGlobalScopeBase::reportViolationForUnsafeEval):
(WebCore::JSWorkerGlobalScopeBase::codeForEval): Deleted.
(WebCore::JSWorkerGlobalScopeBase::canCompileStrings): Deleted.
* Source/WebCore/bindings/js/JSWorkerGlobalScopeBase.h:
* Source/WebCore/bindings/js/JSWorkletGlobalScopeBase.cpp:
(WebCore::JSWorkletGlobalScopeBase::globalObjectMethodTable):
(WebCore::JSWorkletGlobalScopeBase::reportViolationForUnsafeEval):
(WebCore::JSWorkletGlobalScopeBase::codeForEval): Deleted.
(WebCore::JSWorkletGlobalScopeBase::canCompileStrings): Deleted.
* Source/WebCore/bindings/js/JSWorkletGlobalScopeBase.h:
* Source/WebCore/dom/TrustedType.cpp:
(WebCore::canCompile):
* Source/WebCore/dom/TrustedType.h:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations: Remove TrustedType.cpp because the file has been fixed.

Canonical link: <a href="https://commits.webkit.org/287909@main">https://commits.webkit.org/287909@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b844e0180ed9190e3e293cf040b286b2a6028632

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81255 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/780 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35198 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85784 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32241 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83365 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/798 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8594 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63442 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21210 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84324 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/543 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73955 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43740 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/442 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28117 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30699 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/74234 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/71940 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28692 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87219 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/80313 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8485 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/6031 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71752 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8664 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69789 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70987 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17676 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15037 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13946 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/102720 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8446 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13970 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24966 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8283 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11803 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10091 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/FunctionConstructor.cpp:
```diff
@@ -145,6 +145,25 @@ JSObject* constructFunction(JSGlobalObject* globalObject, const ArgList& args, c
+        if (args.size() == 1) {
+            auto body = args.at(0);
+            RETURN_IF_EXCEPTION(scope, { });
+            if (!globalObject->globalObjectMethodTable()->canCompileStrings(globalObject, CompilationType::Function, code, args, body)) {
```
This is too costly. `Function` constructor is actually hot function and we should not call this for the major path.
Can you instead set the information in JSGlobalObject and avoid calling out to WebCore each time we call Function constructor?

lukewarlow replied:
What information could I put into JSObject, I ultimately have to be able to call code from WebCore on the arguments to Function?

lukewarlow replied:
To follow up on this, the code needs to be able to execute a user defined callback for each call of the Function constructor, this callback will be unique per realm (worker and window can define their own).

One possible perf improvement is to store if the necessary CSP is configured and then only run this code when TT is present (Whether it's report only or not). That's probably the best I can do?

Once the PRs this one depends on are merged this code will be gated by a JSC option anyway but I will seak to address any perf issues regardless.

lukewarlow replied:
I've updated this so it no longer calls out to WebCore *unless* you have trusted types enforced in CSP (and have the JSC option enabled)

Constellation reviewed (changes requested):
Commented.

justinmichaud commented on Source/JavaScriptCore/runtime/FunctionConstructor.cpp:
```diff
@@ -145,6 +145,11 @@ JSObject* constructFunction(JSGlobalObject* globalObject, const ArgList& args, c
     auto code = stringifyFunction(globalObject, args, functionName, functionConstructionMode, scope, functionConstructorParametersEndPosition);
     EXCEPTION_ASSERT(!!scope.exception() == code.isNull());
 
+    if (Options::useTrustedTypes() && globalObject->requiresTrustedTypes() && !globalObject->globalObjectMethodTable()->canCompileStrings(globalObject, CompilationType::Function, code, args, jsUndefined())) {
```
Once this option is enabled by default, this will still be a perf regression, right? Is there a way we can chache these canCompileString results on the global object?

lukewarlow replied:
It will only regress when TT is enforced through CSP. We probably can't cache either because it can end up executing JS callbacks which could be strateful (no guarantee it will return the same result each call).

Constellation replied:
When is this `canCompileStrings` used? Is it super rare?
Function constructor is very hot code. And if it is not rare, you should make most of part inlined in JSC side and make it almost zero-cost. If this "requiresTrustedTypes" condition is super rare in the wild, then it is fine.

Constellation reviewed (changes requested):
Discussed in Slack. Right now, this is very naively going to DOM world from Function constructor, and as we said, Function constructor is hot. We should not do that at least when no TT callback is set.

justinmichaud commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -124,12 +124,19 @@ JSValue eval(CallFrame* callFrame, JSValue thisValue, JSScope* callerScopeChain,
-            programString = jsString(vm, code);
+        auto* structure = globalObject->trustedScriptStructure();
+        if (!structure) {
+            structure = globalObject->globalObjectMethodTable()->trustedScriptStructure(globalObject);
```
This should have an UNLIKELY()

@Constellation  Is it faster to do this in the global object constructor instead?

lukewarlow replied:
I've added the UNLIKELY

Constellation replied:
Yeah, do that in global object constructor instead.

justinmichaud commented on Source/JavaScriptCore/runtime/FunctionConstructor.cpp:
```diff
@@ -146,6 +146,28 @@ JSObject* constructFunction(JSGlobalObject* globalObject, const ArgList& args, c
+        bool isTrusted = true;
+        auto* structure = globalObject->trustedScriptStructure();
+        if (!structure) {
+            structure = globalObject->globalObjectMethodTable()->trustedScriptStructure(globalObject);
```
I think it is fine if we can't freeze the structure, as long as the slow canCompileStrings method can handle the case when you pass in a modified trusted script object too. If we discover that it is common to polyfill this type in the future, then we can replace this with a flag on structure.

lukewarlow replied:
Right now it won't allow you to eval a modified trusted script object.
I could keep codeForEval as a slow path but that would make eval of any object slower, which seems bad.

If we can freeze the object during creation that solves that problem and that feels like the correct API design to me.

justinmichaud reviewed (commented):
LGTM, with some extra UNLIKELYs sprinkled in

justinmichaud commented on Source/WebCore/bindings/js/JSDOMGlobalObject.cpp:
```diff
@@ -390,6 +391,13 @@ bool JSDOMGlobalObject::canCompileStrings(JSGlobalObject* globalObject, Compilat
+{
+    auto& thisObject = static_cast<JSDOMGlobalObject&>(*globalObject);
+
+    return getDOMStructure<JSTrustedScript>(globalObject->vm(), thisObject);
```
It seems that the dom wrapper does not keep this alive

justinmichaud replied:
We should either add a WriteBarrier<> and mark it in visitChildrenImpl, or make it a weakref

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObject.h:
```diff
@@ -568,6 +568,7 @@ class JSGlobalObject : public JSSegmentedVariableObject {
     bool m_requiresTrustedTypes { true };
     bool m_needsSiteSpecificQuirks { false };
     unsigned m_globalLexicalBindingEpoch { 1 };
+    WriteBarrier<Structure> m_trustedScriptStructure;
```
Use `WriteBarrierStructureID`, and place it in the fields of JSGlobalObject which has many WriteBarrierStructureID.

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObject.h:
```diff
@@ -1060,6 +1062,10 @@ class JSGlobalObject : public JSSegmentedVariableObject {
+    void setTrustedScriptStructure(Structure* structure)
+    {
+        m_trustedScriptStructure.set(vm(), this, structure);
+    }
```
If we do configuring this in global object constructor, we do not need this function.

Constellation commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -124,12 +124,21 @@ JSValue eval(CallFrame* callFrame, JSValue thisValue, JSScope* callerScopeChain,
-            programString = jsString(vm, code);
+    else if (Options::useTrustedTypes() && program.isObject()) {
+        auto* structure = globalObject->trustedScriptStructure();
+        if (structure == program.structureOrNull()) {
```
Why is it `structureOrNull`? If program is object, then it must have structure.

Constellation commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -124,12 +124,21 @@ JSValue eval(CallFrame* callFrame, JSValue thisValue, JSScope* callerScopeChain,
+    else if (Options::useTrustedTypes() && program.isObject()) {
+        auto* structure = globalObject->trustedScriptStructure();
+        if (structure == program.structureOrNull()) {
+            programString = program.toStringOrNull(globalObject);
```
This is not having exception check

Constellation commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -124,12 +124,21 @@ JSValue eval(CallFrame* callFrame, JSValue thisValue, JSScope* callerScopeChain,
+            programString = program.toStringOrNull(globalObject);
+            isTrusted = true;
+        } else {
+            auto code = globalObject->globalObjectMethodTable()->codeForEval(globalObject, program);
```
This is not having exception check.

Constellation commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -124,12 +124,21 @@ JSValue eval(CallFrame* callFrame, JSValue thisValue, JSScope* callerScopeChain,
+        } else {
+            auto code = globalObject->globalObjectMethodTable()->codeForEval(globalObject, program);
+            if (!code.isNull()) {
+                programString = jsString(vm, code);
```
Why do we need to wrap it with string? This is unnecessary allocation.

lukewarlow replied:
Because it's just a `String` wrapping in jsString because programString is a JSString?

Constellation replied:
I mean, there is very high possibility that this input `code` is not changed as a result. So we should avoid allocating new string here for the major path. We normally do not care about this, but `eval` etc. is hot path, and very small changes easily affect on the real world performance.

lukewarlow replied:
What would you recommend we do instead?

lukewarlow replied:
@Constellation just coming back to this and I've realised this isn't the hot path. The structure check is the hot path. This is for when they're evaling non trusted objects OR where something unusual has happened to the trusted object such that the structure match fails.

Given that I don't think we have to worry about these allocations as much? Unless you're saying that eval of generic objects is a hotpath?

If this does indeed require code changes, I'm also unsure what you mean by `there is very high possibility that this input code is not changed as a result` code is a String and we need a JSString at this point in the code flow? Do you have a code snippet example of what you'd like this to look like?

Constellation replied:
Why do we need to allocate this string? I think you can just get `String programSource` early, changing `reportViolationForUnsafeEval` to take `const String&` instead of `JSString*` is enough, right?

lukewarlow replied:
I wanted to avoid doing too much of a refactor unrelated to the code I'm trying to add. But I've updated the code as suggested.

Constellation commented on Source/JavaScriptCore/runtime/FunctionConstructor.cpp:
```diff
@@ -146,6 +146,24 @@ JSObject* constructFunction(JSGlobalObject* globalObject, const ArgList& args, c
+        for (size_t i = 0; i < args.size(); i++) {
+            auto arg = args.at(i);
+
+            if (!arg.isObject() || structure != arg.structureOrNull()) {
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/FunctionConstructor.cpp:
```diff
@@ -146,6 +146,24 @@ JSObject* constructFunction(JSGlobalObject* globalObject, const ArgList& args, c
+            }
+        }
+
+        if (!isTrusted && !globalObject->globalObjectMethodTable()->canCompileStrings(globalObject, CompilationType::Function, code, args)) {
```
This does not have exception check.

Constellation replied:
Ditto.

lukewarlow replied:
I've added a RETURN_IF_EXCEPTION inside of and after the if statement to catch both branches.

Constellation replied:
That's not correct. Regardless of whether the exception is thrown, we need to check, so this code will gets crash with assertion.
You need to make it like,
```
if (!isTrusted) {
    bool canCompileStrings = globalObject->globalObjectMethodTable()->canCompileStrings(globalObject, CompilationType::Function, code, args);
    RETURN_IF_EXCEPTION(scope, { });
    ...
```

Please ensure that all of the added code is following to the above pattern precisely if it can throw an error / it is using exception scope.

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
```diff
@@ -477,12 +477,21 @@ JSC_DEFINE_HOST_FUNCTION(globalFuncEval, (JSGlobalObject* globalObject, CallFram
-        if (!code.isNull())
-            codeString = jsString(vm, code);
+        auto* structure = globalObject->trustedScriptStructure();
+        if (structure == x.structureOrNull()) {
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
```diff
@@ -477,12 +477,21 @@ JSC_DEFINE_HOST_FUNCTION(globalFuncEval, (JSGlobalObject* globalObject, CallFram
-            codeString = jsString(vm, code);
+        auto* structure = globalObject->trustedScriptStructure();
+        if (structure == x.structureOrNull()) {
+            codeString = x.toStringOrNull(globalObject);
```
This does not have exception check.

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
```diff
@@ -477,12 +477,21 @@ JSC_DEFINE_HOST_FUNCTION(globalFuncEval, (JSGlobalObject* globalObject, CallFram
+            codeString = x.toStringOrNull(globalObject);
+            isTrusted = true;
+        } else {
+            auto code = globalObject->globalObjectMethodTable()->codeForEval(globalObject, x);
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
```diff
@@ -477,12 +477,21 @@ JSC_DEFINE_HOST_FUNCTION(globalFuncEval, (JSGlobalObject* globalObject, CallFram
+        } else {
+            auto code = globalObject->globalObjectMethodTable()->codeForEval(globalObject, x);
+            if (!code.isNull()) {
+                codeString = jsString(vm, code);
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
```diff
@@ -491,9 +500,9 @@ JSC_DEFINE_HOST_FUNCTION(globalFuncEval, (JSGlobalObject* globalObject, CallFram
     RETURN_IF_EXCEPTION(scope, encodedJSValue());
 
-    if (Options::useTrustedTypes() && globalObject->requiresTrustedTypes() && !globalObject->globalObjectMethodTable()->canCompileStrings(globalObject, CompilationType::IndirectEval, s, x)) {
+    if (Options::useTrustedTypes() && globalObject->requiresTrustedTypes() && !isTrusted && !globalObject->globalObjectMethodTable()->canCompileStrings(globalObject, CompilationType::IndirectEval, s, *vm.emptyList)) {
```
Ditto.

lukewarlow replied:
Hopefully this has been addressed properly

Constellation replied:
Ditto. This is not correct.

Constellation commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -138,7 +147,7 @@ JSValue eval(CallFrame* callFrame, JSValue thisValue, JSScope* callerScopeChain,
     RETURN_IF_EXCEPTION(scope, JSValue());
 
-    if (Options::useTrustedTypes() && globalObject->requiresTrustedTypes() && !globalObject->globalObjectMethodTable()->canCompileStrings(globalObject, CompilationType::DirectEval, programSource, program)) {
+    if (Options::useTrustedTypes() && globalObject->requiresTrustedTypes() && !isTrusted && !globalObject->globalObjectMethodTable()->canCompileStrings(globalObject, CompilationType::DirectEval, programSource, *vm.emptyList)) {
```
This `canCompileStrings` does not have exception check.

lukewarlow replied:
If i put a RETURN_IF_EXCEPTION isnide of the if block it fails the tests because it starts returning a typeerror rather than the required eval.

I have put a `scope.clearException();` inside instead based on similar code in the unsafe-eval check in FunctionConstructor.cpp

Constellation replied:
Hmm, this sounds strange. Swallowing the already thrown error is not a good thing.
Why is TypeError thrown from the first place? Can you replace that TypeError with EvalError when it is necessary?

lukewarlow replied:
The default policy can also end up throwing arbitrary exception types. Normally these TypeErrors and arbitrary userland exceptions would be propogated up but for eval and Function they all need to be swallowed and replaced with an EvalError.

I'm happy to move the scope.clearExceptions down into canCompileStrings and return a false for when an exception exists rather than propagating up to here to be cleared?

Constellation replied:
Can you specify the part of the spec corresponding to this part?

> The default policy can also end up throwing arbitrary exception types. Normally these TypeErrors and arbitrary userland exceptions would be propogated up but for eval and Function they all need to be swallowed and replaced with an EvalError.

Constellation replied:
Also, regardless, we need exception check after calling `canCompileStrings`, like, `exception()` etc.
JSC is checking exception check is done after leaving exception scope (it exists inside `canCompileStrings`), and crash with debug assertion. So with the current code, JSC will crash with debug assert.

lukewarlow replied:
I've rejigged the code a bit and added the RETURN_IF_EXCEPTION checks inside of the if statement and afterwards to catch both code branches

Constellation replied:
Ditto.

Constellation commented on Source/WebCore/bindings/js/JSDOMWindowBase.cpp:
```diff
@@ -304,17 +305,17 @@ void JSDOMWindowBase::reportViolationForUnsafeEval(JSGlobalObject* object, JSStr
-
-    if (auto* script = JSTrustedScript::toWrapped(vm, value))
-        return script->toString();
+    return JSDOMGlobalObject::codeForEval(globalObject, value);
```
Why do we need to wrap this while JSDOMWindowBase's base class is JSDOMGlobalObject?

lukewarlow replied:
I was following the pattern of the surrounding code but you're right I don't need to do any of this stuff it just works with the definitions in JSDOMGlobalObject. Will clean that up.

Constellation commented on Source/WebCore/bindings/js/JSDOMWindowBase.cpp:
```diff
@@ -304,17 +305,17 @@ void JSDOMWindowBase::reportViolationForUnsafeEval(JSGlobalObject* object, JSStr
-    return nullString();
+bool JSDOMWindowBase::canCompileStrings(JSGlobalObject* globalObject, CompilationType compilationType, String codeString, const ArgList& args)
+{
+    return JSDOMGlobalObject::canCompileStrings(globalObject, compilationType, codeString, args);
```
Ditto.

Constellation commented on Source/WebCore/bindings/js/JSDOMWindowBase.cpp:
```diff
@@ -304,17 +305,17 @@ void JSDOMWindowBase::reportViolationForUnsafeEval(JSGlobalObject* object, JSStr
+Structure* JSDOMWindowBase::trustedScriptStructure(JSGlobalObject* globalObject)
 {
-    return JSDOMGlobalObject::canCompileStrings(globalObject, compilationType, codeString, bodyArgument);
+    return JSDOMGlobalObject::trustedScriptStructure(globalObject);
```
Ditto.

Constellation commented on Source/WebCore/bindings/js/JSShadowRealmGlobalScopeBase.cpp:
```diff
@@ -160,9 +161,14 @@ String JSShadowRealmGlobalScopeBase::codeForEval(JSC::JSGlobalObject* globalObje
+bool JSShadowRealmGlobalScopeBase::canCompileStrings(JSC::JSGlobalObject* globalObject, JSC::CompilationType compilationType, String codeString, const ArgList& args)
 {
-    return JSGlobalObject::canCompileStrings(globalObject, compilationType, codeString, bodyArgument);
+    return JSGlobalObject::canCompileStrings(globalObject, compilationType, codeString, args);
```
Ditto.

Constellation commented on Source/WebCore/bindings/js/JSShadowRealmGlobalScopeBase.cpp:
```diff
@@ -160,9 +161,14 @@ String JSShadowRealmGlobalScopeBase::codeForEval(JSC::JSGlobalObject* globalObje
+
+JSC::Structure* JSShadowRealmGlobalScopeBase::trustedScriptStructure(JSC::JSGlobalObject* globalObject)
+{
+    return JSGlobalObject::trustedScriptStructure(globalObject);
```
Ditto.

Constellation commented on Source/WebCore/bindings/js/JSWorkerGlobalScopeBase.cpp:
```diff
@@ -151,17 +152,17 @@ void JSWorkerGlobalScopeBase::reportViolationForUnsafeEval(JSC::JSGlobalObject*
-
-    if (auto* script = JSTrustedScript::toWrapped(vm, value))
-        return script->toString();
+    return JSDOMGlobalObject::codeForEval(globalObject, value);
```
Ditto.

Constellation commented on Source/WebCore/bindings/js/JSWorkerGlobalScopeBase.cpp:
```diff
@@ -151,17 +152,17 @@ void JSWorkerGlobalScopeBase::reportViolationForUnsafeEval(JSC::JSGlobalObject*
-    return nullString();
+bool JSWorkerGlobalScopeBase::canCompileStrings(JSC::JSGlobalObject* globalObject, JSC::CompilationType compilationType, String codeString, const ArgList& args)
+{
+    return JSDOMGlobalObject::canCompileStrings(globalObject, compilationType, codeString, args);
```
Ditto.

Constellation commented on Source/WebCore/bindings/js/JSWorkerGlobalScopeBase.cpp:
```diff
@@ -151,17 +152,17 @@ void JSWorkerGlobalScopeBase::reportViolationForUnsafeEval(JSC::JSGlobalObject*
+JSC::Structure* JSWorkerGlobalScopeBase::trustedScriptStructure(JSC::JSGlobalObject* globalObject)
 {
-    return JSDOMGlobalObject::canCompileStrings(globalObject, compilationType, codeString, bodyArgument);
+    return JSDOMGlobalObject::trustedScriptStructure(globalObject);
```
Ditto.

Constellation commented on Source/WebCore/bindings/js/JSWorkletGlobalScopeBase.cpp:
```diff
@@ -126,9 +127,14 @@ String JSWorkletGlobalScopeBase::codeForEval(JSC::JSGlobalObject* globalObject,
+bool JSWorkletGlobalScopeBase::canCompileStrings(JSC::JSGlobalObject* globalObject, JSC::CompilationType compilationType, String codeString, const ArgList& args)
 {
-    return JSGlobalObject::canCompileStrings(globalObject, compilationType, codeString, bodyArgument);
+    return JSGlobalObject::canCompileStrings(globalObject, compilationType, codeString, args);
```
It is calling `JSGlobalObject` instead of JSDOMGlobalObject, is it right?

Constellation commented on Source/WebCore/bindings/js/JSWorkletGlobalScopeBase.cpp:
```diff
@@ -126,9 +127,14 @@ String JSWorkletGlobalScopeBase::codeForEval(JSC::JSGlobalObject* globalObject,
+
+JSC::Structure* JSWorkletGlobalScopeBase::trustedScriptStructure(JSC::JSGlobalObject* globalObject)
+{
+    return JSGlobalObject::trustedScriptStructure(globalObject);
```
Ditto, is it right?

Constellation commented on Source/WebCore/dom/TrustedType.cpp:
```diff
@@ -302,14 +303,33 @@ ExceptionOr<String> requireTrustedTypesForPreNavigationCheckPasses(ScriptExecuti
+                break;
+            }
+            if (auto trustedScript = JSTrustedScript::toWrapped(vm, args.at(i))) {
+                if (!trustedScript || trustedScript->toString() != arg.toWTFString(scriptExecutionContext.globalObject())) {
```
toWTFString is not having exception check.

Constellation reviewed (changes requested):
OK, commented.

Constellation commented on Source/WebCore/dom/TrustedType.cpp:
```diff
@@ -302,14 +303,35 @@ ExceptionOr<String> requireTrustedTypesForPreNavigationCheckPasses(ScriptExecuti
+            }
+            if (auto trustedScript = JSTrustedScript::toWrapped(vm, arg)) {
+                if (!trustedScript || trustedScript->toString() != arg.toWTFString(scriptExecutionContext.globalObject())) {
+                    RETURN_IF_EXCEPTION(scope, Exception { ExceptionCode::ExistingExceptionError });
```
is is not a correct place to emit exception check. When the above code pass with == then no exception check exist.
Every time we run `toWTFString` etc., just after that, exception check should be done.

lukewarlow replied:
I've added another one after the if statement is that correct?

Constellation replied:
That's not great. And we are using string returned from toWTFString without exception check. You should not put exception check when it is not necessary. Please reorganize,
```
if (!trustedScript) {
    isTrusted = false;
    break;
}
auto argString = arg.toWTFString(scriptExecutionContext.globalObject());
RETURN_IF_EXCEPTION(scope, Exception { ExceptionCode::ExistingExceptionError });
if (trustedScript->toString() != argString) {
    isTrusted = false;
    break;
}
```

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
```diff
@@ -476,38 +476,49 @@ JSC_DEFINE_HOST_FUNCTION(globalFuncEval, (JSGlobalObject* globalObject, CallFram
         throwException(globalObject, scope, createEvalError(globalObject, "Refused to evaluate a string as JavaScript because this document requires a 'Trusted Type' assignment."_s));
         return { };
     }
+    RETURN_IF_EXCEPTION(scope, encodedJSValue());
```
Let's not do this. We should follow to the right idiom. Like,
```
bool canCompileString = globalObject->globalObjectMethodTable()->canCompileStrings(globalObject, CompilationType::IndirectEval, s, *vm.emptyList);
RETURN_IF_EXCEPTION(scope, encodedJSValue());
...
```

lukewarlow replied:
Fixed. Sorry for the delay.

fred-wang commented on Source/WebCore/bindings/js/JSDOMGlobalObject.h:
```diff
@@ -79,7 +79,9 @@ class WEBCORE_EXPORT JSDOMGlobalObject : public JSC::JSGlobalObject {
     ScriptExecutionContext* scriptExecutionContext() const;
 
-    static bool canCompileStrings(JSC::JSGlobalObject*, JSC::CompilationType, String, JSC::JSValue);
+    static String codeForEval(JSC::JSGlobalObject*, JSC::JSValue);
```
Maybe add a link to https://html.spec.whatwg.org/multipage/webappapis.html#hostgetcodeforeval(argument) ?

fred-wang commented on Source/WebCore/bindings/js/JSDOMGlobalObject.h:
```diff
@@ -79,7 +79,9 @@ class WEBCORE_EXPORT JSDOMGlobalObject : public JSC::JSGlobalObject {
 
-    static bool canCompileStrings(JSC::JSGlobalObject*, JSC::CompilationType, String, JSC::JSValue);
+    static String codeForEval(JSC::JSGlobalObject*, JSC::JSValue);
+    static bool canCompileStrings(JSC::JSGlobalObject*, JSC::CompilationType, String, const JSC::ArgList&);
```
and to https://html.spec.whatwg.org/multipage/webappapis.html#hostensurecancompilestrings(realm,-parameterstrings,-bodystring,-codestring,-compilationtype,-parameterargs,-bodyarg) ?

lukewarlow replied:
I don't think it's common practice to add links or comments in general to headers in webkit.

These functions are also not necessarily a 1-1 matchup to the specs because of perf optimisations on the JSC side so might be more confusing than helpful?

fred-wang commented on Source/WebCore/bindings/js/JSShadowRealmGlobalScopeBase.cpp:
```diff
@@ -149,22 +150,12 @@ JSC::ScriptExecutionStatus JSShadowRealmGlobalScopeBase::scriptExecutionStatus(J
-{
-    return JSGlobalObject::canCompileStrings(globalObject, compilationType, codeString, bodyArgument);
-}
-
```
The ShadowRealm does not say anything for now: https://github.com/tc39/proposal-shadowrealm/issues/414 and I don't think this is covered by any test. Shouldn't we keep current behavior for now?

lukewarlow replied:
This just ends up going through the parent implementation so should be fine.

fred-wang commented on Source/WebCore/dom/TrustedType.cpp:
```diff
@@ -302,12 +303,36 @@ ExceptionOr<String> requireTrustedTypesForPreNavigationCheckPasses(ScriptExecuti
 }
 
-ExceptionOr<bool> canCompile(ScriptExecutionContext& scriptExecutionContext, JSC::CompilationType compilationType, String codeString, JSC::JSValue bodyArgument)
+ExceptionOr<bool> canCompile(ScriptExecutionContext& scriptExecutionContext, JSC::CompilationType compilationType, String codeString, const JSC::ArgList& args)
```
FYI, more tests for that have landed at https://github.com/web-platform-tests/wpt/commit/1076fe42f0ea43b6b1ea3b7bb13307bbee5cfafb

fred-wang commented on Source/WebCore/dom/TrustedType.cpp:
```diff
@@ -302,12 +303,36 @@ ExceptionOr<String> requireTrustedTypesForPreNavigationCheckPasses(ScriptExecuti
 {
-    if (bodyArgument.isObject())
-        return JSTrustedScript::toWrapped(scriptExecutionContext.vm(), bodyArgument) ? true : false;
+    if (compilationType == CompilationType::Function) {
```
The spec says we should do that for DirectEval and IndirectEval. However, in that case this really simplifies to `if (args.at(0).isObject()) return true;`

See the MOZ_ASSERT I have in my Gecko patch: https://phabricator.services.mozilla.com/D229624

Isn't this covered by existing tests?

lukewarlow replied:
This is covered in tests I'm pretty sure, it's just the model here can't directly match the spec flow for various reasons.

fred-wang commented on LayoutTests/imported/w3c/web-platform-tests/trusted-types/default-policy-callback-arguments.html:
```diff
@@ -47,7 +47,7 @@
           _ => script.src = "about:blank" ],
       [ "2+2", "TrustedScript", "eval", _ => eval("2+2") ],
-      [ "(function anonymous(\n) {\nreturn 2+2\n})", "TrustedScript",
+      [ "function anonymous(\n) {\nreturn 2+2\n}", "TrustedScript",
```
Just for the record: this expectation seems to be based on Chromium's behavior, but indeed https://tc39.es/ecma262/#sec-createdynamicfunction does not add parentheses around *sourceString*

Constellation commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -123,29 +123,42 @@ JSValue eval(CallFrame* callFrame, JSValue thisValue, JSScope* callerScopeChain,
+        auto* structure = globalObject->trustedScriptStructure();
+        if (structure == asObject(program)->structure()) {
+            programSource = program.toWTFString(globalObject);
+            RETURN_IF_EXCEPTION(scope, JSValue());
```
This should be `RETURN_IF_EXCEPTION(scope, { });`

Constellation commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -123,29 +123,42 @@ JSValue eval(CallFrame* callFrame, JSValue thisValue, JSScope* callerScopeChain,
+    bool isTrusted = false;
+    if (LIKELY(program.isString())) {
+        programSource = program.toWTFString(globalObject);
+        RETURN_IF_EXCEPTION(scope, JSValue());
```
This should be `RETURN_IF_EXCEPTION(scope, { });`

Constellation commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -123,29 +123,42 @@ JSValue eval(CallFrame* callFrame, JSValue thisValue, JSScope* callerScopeChain,
+            isTrusted = true;
+        } else {
+            auto code = globalObject->globalObjectMethodTable()->codeForEval(globalObject, program);
+            RETURN_IF_EXCEPTION(scope, JSValue());
```
This should be `RETURN_IF_EXCEPTION(scope, { });`

Constellation commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -123,29 +123,42 @@ JSValue eval(CallFrame* callFrame, JSValue thisValue, JSScope* callerScopeChain,
-        auto code = globalObject->globalObjectMethodTable()->codeForEval(globalObject, program);
-        if (!code.isNull())
-            programString = jsString(vm, code);
+    String programSource = nullString();
```
You can just use `String programSource;`

Constellation commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -123,29 +123,42 @@ JSValue eval(CallFrame* callFrame, JSValue thisValue, JSScope* callerScopeChain,
-        return { };
+    if (Options::useTrustedTypes() && globalObject->requiresTrustedTypes() && !isTrusted) {
+        bool canCompileStrings = globalObject->globalObjectMethodTable()->canCompileStrings(globalObject, CompilationType::DirectEval, programSource, *vm.emptyList);
+        RETURN_IF_EXCEPTION(scope, JSValue());
```
This should be `RETURN_IF_EXCEPTION(scope, { });`

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
```diff
@@ -478,38 +478,51 @@ JSC_DEFINE_HOST_FUNCTION(globalFuncEval, (JSGlobalObject* globalObject, CallFram
-        auto code = globalObject->globalObjectMethodTable()->codeForEval(globalObject, x);
-        if (!code.isNull())
-            codeString = jsString(vm, code);
+    String s = nullString();
```
Ditto. And let's not use `s`, it is too hard to follow. Use descriptive name.

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
```diff
@@ -478,38 +478,51 @@ JSC_DEFINE_HOST_FUNCTION(globalFuncEval, (JSGlobalObject* globalObject, CallFram
+    bool isTrusted = false;
+    if (LIKELY(x.isString())) {
+        s = x.toWTFString(globalObject);
+        RETURN_IF_EXCEPTION(scope, encodedJSValue());
```
This should be `RETURN_IF_EXCEPTION(scope, { });`

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
```diff
@@ -478,38 +478,51 @@ JSC_DEFINE_HOST_FUNCTION(globalFuncEval, (JSGlobalObject* globalObject, CallFram
+        auto* structure = globalObject->trustedScriptStructure();
+        if (structure == asObject(x)->structure()) {
+            s = x.toWTFString(globalObject);
+            RETURN_IF_EXCEPTION(scope, encodedJSValue());
```
This should be `RETURN_IF_EXCEPTION(scope, { });`

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
```diff
@@ -478,38 +478,51 @@ JSC_DEFINE_HOST_FUNCTION(globalFuncEval, (JSGlobalObject* globalObject, CallFram
+            isTrusted = true;
+        } else {
+            auto code = globalObject->globalObjectMethodTable()->codeForEval(globalObject, x);
+            RETURN_IF_EXCEPTION(scope, encodedJSValue());
```
This should be `RETURN_IF_EXCEPTION(scope, { });`

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
```diff
@@ -478,38 +478,51 @@ JSC_DEFINE_HOST_FUNCTION(globalFuncEval, (JSGlobalObject* globalObject, CallFram
-        return { };
+    if (Options::useTrustedTypes() && globalObject->requiresTrustedTypes() && !isTrusted) {
+        bool canCompileStrings = globalObject->globalObjectMethodTable()->canCompileStrings(globalObject, CompilationType::IndirectEval, s, *vm.emptyList);
+        RETURN_IF_EXCEPTION(scope, encodedJSValue());
```
This should be `RETURN_IF_EXCEPTION(scope, { });`

Constellation commented on Source/WebCore/bindings/js/JSDOMGlobalObject.cpp:
```diff
@@ -372,24 +373,45 @@ ScriptExecutionContext* JSDOMGlobalObject::scriptExecutionContext() const
+    if (auto* script = JSTrustedScript::toWrapped(vm, value))
+        return script->toString();
+
+    return nullString();
```
You can just use `return String();`

Constellation commented on Source/WebCore/bindings/js/JSDOMGlobalObject.cpp:
```diff
@@ -372,24 +373,45 @@ ScriptExecutionContext* JSDOMGlobalObject::scriptExecutionContext() const
+        // Step 2.7. If the algorithm throws an error, throw an EvalError.
+        // This clears the existing exceptions and returns false, where the caller throws an EvalError.
+        throwScope.clearException();
         RETURN_IF_EXCEPTION(throwScope, false);
```
We do not need this line since we cleared the exception.

lukewarlow committed:
Implement trusted types enforcement on Function constructor
https://bugs.webkit.org/show_bug.cgi?id=273187

Reviewed by Yusuke Suzuki.

This patch adds trusted types enforcement to the Function constructor as well as updating the eval implementation.

The canCompileStrings global method table function no longer takes a JSValue argument,
this is because it is now only called for untrusted input from eval.

The implementation of TT enforcement for both direct and indirect eval is updated to do more work in JSC.
The structure of the TrustedScript type is used by JSC to determine if an object should be evaluated,
rather than always calling codeForEval. Only if the structures don't match is codeForEval called, this can happen
if someone changes the instance properties such as for polyfills.

The canCompileStrings call is now only done if the input is known
to be untrusted (raw string rather than TrustedScript argument).

The Function constructor is now also updated such that when TT enforcement is enabled through CSP, similar logic
comparing the arguments to the TrustedScript structure is used.
If not all of the arguments match the structure then they're not trusted so we fallback to calling canCompileStrings,
with a new ArgList atgument.

The ArgList is used to check if the arguments are modified trusted script objects, which are accepted providing the
stringifier isn't modified.

DebuggerEvalEnabler is also updated to disable trusted types and re-enable it, so that
web inspector can continue working on sites with TT enforced.

* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/reporting/report-clips-sample.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-csp-tt-default-policy-mutate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-csp-tt-no-default-policy-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-function-constructor-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-function-constructor.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-with-permissive-csp-expected.txt:
* Source/JavaScriptCore/API/JSAPIGlobalObject.cpp:
(JSC::JSAPIGlobalObject::globalObjectMethodTable):
* Source/JavaScriptCore/API/JSAPIGlobalObject.mm:
(JSC::JSAPIGlobalObject::globalObjectMethodTable):
* Source/JavaScriptCore/debugger/DebuggerEvalEnabler.h:
(JSC::DebuggerEvalEnabler::DebuggerEvalEnabler):
(JSC::DebuggerEvalEnabler::~DebuggerEvalEnabler):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::eval):
* Source/JavaScriptCore/jsc.cpp:
* Source/JavaScriptCore/runtime/DirectEvalExecutable.cpp:
(JSC::DirectEvalExecutable::create):
* Source/JavaScriptCore/runtime/FunctionConstructor.cpp:
(JSC::constructFunction):
* Source/JavaScriptCore/runtime/GlobalObjectMethodTable.h:
* Source/JavaScriptCore/runtime/IndirectEvalExecutable.cpp:
(JSC::IndirectEvalExecutable::createImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::baseGlobalObjectMethodTable):
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::arrayIteratorStructure const):
(JSC::JSGlobalObject::trustedScriptStructure):
(JSC::JSGlobalObject::reportViolationForUnsafeEval):
(JSC::JSGlobalObject::canCompileStrings):
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WebCore/bindings/js/JSDOMGlobalObject.cpp:
(WebCore::JSDOMGlobalObject::codeForEval):
(WebCore::JSDOMGlobalObject::canCompileStrings):
(WebCore::JSDOMGlobalObject::trustedScriptStructure):
* Source/WebCore/bindings/js/JSDOMGlobalObject.h:
* Source/WebCore/bindings/js/JSDOMWindowBase.cpp:
(WebCore::JSDOMWindowBase::globalObjectMethodTable):
(WebCore::JSDOMWindowBase::reportViolationForUnsafeEval):
(WebCore::JSDOMWindowBase::codeForEval): Deleted.
(WebCore::JSDOMWindowBase::canCompileStrings): Deleted.
* Source/WebCore/bindings/js/JSDOMWindowBase.h:
* Source/WebCore/bindings/js/JSShadowRealmGlobalScopeBase.cpp:
(WebCore::JSShadowRealmGlobalScopeBase::globalObjectMethodTable):
(WebCore::JSShadowRealmGlobalScopeBase::reportViolationForUnsafeEval):
(WebCore::JSShadowRealmGlobalScopeBase::codeForEval): Deleted.
(WebCore::JSShadowRealmGlobalScopeBase::canCompileStrings): Deleted.
* Source/WebCore/bindings/js/JSShadowRealmGlobalScopeBase.h:
* Source/WebCore/bindings/js/JSWorkerGlobalScopeBase.cpp:
(WebCore::JSWorkerGlobalScopeBase::globalObjectMethodTable):
(WebCore::JSWorkerGlobalScopeBase::reportViolationForUnsafeEval):
(WebCore::JSWorkerGlobalScopeBase::codeForEval): Deleted.
(WebCore::JSWorkerGlobalScopeBase::canCompileStrings): Deleted.
* Source/WebCore/bindings/js/JSWorkerGlobalScopeBase.h:
* Source/WebCore/bindings/js/JSWorkletGlobalScopeBase.cpp:
(WebCore::JSWorkletGlobalScopeBase::globalObjectMethodTable):
(WebCore::JSWorkletGlobalScopeBase::reportViolationForUnsafeEval):
(WebCore::JSWorkletGlobalScopeBase::codeForEval): Deleted.
(WebCore::JSWorkletGlobalScopeBase::canCompileStrings): Deleted.
* Source/WebCore/bindings/js/JSWorkletGlobalScopeBase.h:
* Source/WebCore/dom/TrustedType.cpp:
(WebCore::canCompile):
* Source/WebCore/dom/TrustedType.h:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations: Remove TrustedType.cpp because the file has been fixed.

Canonical link: https://commits.webkit.org/287909@main
